server/src/server.ml

```diff
@@ -14,83 +14,257 @@
    License for the specific language governing permissions and limitations under
    the License. *)
 
+open Catala_utils
+open Server_types
 open Lwt.Syntax
+open Utils
+module SState = Server_state
+
+let ( let*? ) v f =
+  Lwt.bind v @@ function None -> Lwt.return_none | Some x -> f x
+
+let f (_ : Doc_id.t) = assert false
 
 exception ServerError of string
 
-let parse_settings settings =
-  match settings with
-  | `Assoc [("preludes", `List l)] ->
-    List.map
-      (fun s ->
-        match s with
-        | `String f -> f
-        | _ ->
-          raise
-            (ServerError
-               (Format.asprintf
-                  "Expected a path to a prelude file as a string, instead got \
-                   %a"
-                  Yojson.Safe.pp s)))
-      l
-  | _ ->
-    raise
-      (ServerError
-         (Format.asprintf
-            "ChangeConfiguration: received unexpected settings: %a"
-            Yojson.Safe.pp settings))
-
-let preprocess_uri uri =
-  let prefix = "file://" in
-  let prefix_len = String.length prefix in
-  let uri_len = String.length uri in
-  if prefix_len < uri_len && String.sub uri 0 prefix_len = prefix then
-    String.sub uri prefix_len (uri_len - prefix_len)
-  else uri
-
-let mk_prelude _files = []
-let current_thread = ref Lwt.return_unit
-
-let run_with_delay ?(delay = 1.) ~when_ready f =
-  let run_with_delay () =
-    let* () = Lwt_unix.sleep delay in
-    f ()
+let lookup_project ~notify_back doc_id projects =
+  match Projects.find_or_populate_project ~notify_back doc_id projects with
+  | file, project, `Unchanged -> (file, project), None
+  | file, project, `Changed new_projects -> (file, project), Some new_projects
+
+let set_log_level =
+  let open Linol_lwt.TraceValues in
+  function
+  | None | Some Off ->
+    Logs.app (fun m -> m "switching log level to [none]");
+    Logs.set_level None
+  | Some Messages ->
+    Logs.set_level (Some App);
+    Logs.app (fun m -> m "switching log level to [messages]")
+  | Some Compact ->
+    Logs.set_level (Some Info);
+    Logs.app (fun m -> m "switching log level to [compact]")
+  | Some Verbose ->
+    Logs.set_level (Some Debug);
+    Logs.app (fun m -> m "switching log level to [verbose]")
+
+let protect_project_not_found f =
+  Lwt.catch f (function
+    | Projects.Project_not_found -> Lwt.return_unit
+    | e -> Lwt.fail e)
+
+let protect_project_not_found_opt f =
+  Lwt.catch f (function
+    | Projects.Project_not_found -> Lwt.return_none
+    | e -> Lwt.fail e)
+
+let should_ignore (uri : Doc_id.t) =
+  let file = (uri :> File.t) in
+  let p = File.path_to_list file in
+  let b =
+    List.exists (fun s -> String.length s > 0 && (s.[0] = '.' || s.[0] = '_')) p
```

Comment:
Maybe that's a little too general? Should it only ignore `_build` + an ignore-list declared explicitly in `clerk.toml` ?